src/vs/workbench/contrib/debug/common/debugModel.ts

```diff
@@ -1197,24 +1197,46 @@ export class DataBreakpoint extends BaseBreakpoint implements IDataBreakpoint {
 		this.accessTypes = opts.accessTypes;
 		this.accessType = opts.accessType;
 		if (opts.initialSessionData) {
-			this.sessionDataIdForAddr.set(opts.initialSessionData.session, opts.initialSessionData.dataId);
+			this.sessionDataId.set(opts.initialSessionData.session, opts.initialSessionData.dataId);
 		}
 	}
 
 	async toDAP(session: IDebugSession): Promise<DebugProtocol.DataBreakpoint | undefined> {
-		let dataId: string;
-		if (this.src.type === DataBreakpointSetType.Variable) {
-			dataId = this.src.dataId;
-		} else {
-			let sessionDataId = this.sessionDataIdForAddr.get(session);
-			if (!sessionDataId) {
-				sessionDataId = (await session.dataBytesBreakpointInfo(this.src.address, this.src.bytes))?.dataId;
-				if (!sessionDataId) {
-					return undefined;
+		let dataId = this.sessionDataId.get(session);
+		if (!dataId) {
+			if (this.src.type === DataBreakpointSetType.Variable) {
+				dataId = this.src.dataId;
+			} else if (this.src.type === DataBreakpointSetType.Address) {
+				const sessionDataId = (await session.dataBytesBreakpointInfo(this.src.address, this.src.bytes))?.dataId;
+				if (sessionDataId) {
+					this.sessionDataId.set(session, sessionDataId);
+					dataId = sessionDataId;
+				}
+			} else if (this.src.type === DataBreakpointSetType.Expression) {
+				const sessionDataId = (await session.dataBreakpointInfo(this.src.expression))?.dataId;
+				if (sessionDataId) {
+					this.sessionDataId.set(session, sessionDataId);
+					dataId = sessionDataId;
+				}
+			} else {
+				// type === DataBreakpointSetType.Scoped
+				if (this.src.frameId) {
```

Comment:
Any given frame ID is valid only within the context of the current debug stopped state.

Reply:
Understood.